### ISSUE TITLE
[BUG] Duplicate channel close error in the backing image manage related components

### ISSUE THREAD

**Issue by shuo-wu** [longhorn/longhorn#4865]
## Describe the bug (🐛 if you encounter this issue)

There is duplicate channel closing that will lead to the longhorn manager panic. 
For example, when the backing image data source is unreachable for a while and the monitor failed to retry connecting it, the stop channel will be closed both in the sync function and in a separate stop monitoring function.

## To Reproduce

1. Launch a large backing image 
2. During the backing image downloading and the backing image data source pod is running, freeze the process or cut the network for data source pod. Then wait for a while
3. Check the longhorn manager pod. see the error `panic: close of closed channel`

## Expected behavior

The backing image downloading will be marked as failed without any panic

## Log or Support bundle
```
longhorn-manager-bkx9h longhorn-manager time="2022-11-14T01:49:45Z" level=error msg="There's no available disk for replica pvc-0c193aed-19fb-47dc-804d-e41d48094646-r-cd5ed9e9, size 107374182400"
longhorn-manager-bkx9h longhorn-manager time="2022-11-14T01:49:45Z" level=error msg="unable to schedule replica" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=harvester-v7tfx owner=harvester-v7tfx replica=pvc-0c193aed-19fb-47dc-804d-e41d48094646-r-cd5ed9e9 state=attached volume=pvc-0c193aed-19fb-47dc-804d-e41d48094646
longhorn-manager-bkx9h longhorn-manager time="2022-11-14T01:49:48Z" level=error msg="failed to get default-image-5jcqf info from backing image data source server: get failed, err: Get \"http://10.52.1.23:8000/v1/file\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)" backingImageDataSource=default-image-5jcqf controller=longhorn-backing-image-data-source diskUUID=95813962-0cf4-4224-8378-49a0cec8ae93 node=harvester-v7tfx nodeID=harvester-v7tfx parameters="map[url:http://harvester-vm-import-controller.harvester-system.svc:8080/gm-rke-default-disk-0.img]" sourceType=download
longhorn-manager-bkx9h longhorn-manager time="2022-11-14T01:49:48Z" level=warning msg="Stop monitoring since monitor default-image-5jcqf sync reaches the max retry count 10" backingImageDataSource=default-image-5jcqf controller=longhorn-backing-image-data-source diskUUID=95813962-0cf4-4224-8378-49a0cec8ae93 node=harvester-v7tfx nodeID=harvester-v7tfx parameters="map[url:http://harvester-vm-import-controller.harvester-system.svc:8080/gm-rke-default-disk-0.img]" sourceType=download
longhorn-manager-bkx9h longhorn-manager time="2022-11-14T01:49:48Z" level=info msg="Stopping monitoring" backingImageDataSource=default-image-5jcqf controller=longhorn-backing-image-data-source node=harvester-v7tfx
longhorn-manager-bkx9h longhorn-manager panic: close of closed channel
longhorn-manager-bkx9h longhorn-manager 
longhorn-manager-bkx9h longhorn-manager goroutine 3867 [running]:
longhorn-manager-bkx9h longhorn-manager github.com/longhorn/longhorn-manager/controller.(*BackingImageDataSourceController).stopMonitoring(0xc0004c0000, {0xc00370e270, 0x13})
longhorn-manager-bkx9h longhorn-manager         /go/src/github.com/longhorn/longhorn-manager/controller/backing_image_data_source_controller.go:930 +0x145
longhorn-manager-bkx9h longhorn-manager github.com/longhorn/longhorn-manager/controller.(*BackingImageDataSourceController).startMonitoring.func1()
longhorn-manager-bkx9h longhorn-manager         /go/src/github.com/longhorn/longhorn-manager/controller/backing_image_data_source_controller.go:970 +0x4c
longhorn-manager-bkx9h longhorn-manager created by github.com/longhorn/longhorn-manager/controller.(*BackingImageDataSourceController).startMonitoring
longhorn-manager-bkx9h longhorn-manager         /go/src/github.com/longhorn/longhorn-manager/controller/backing_image_data_source_controller.go:968 +0x465
```

## Environment

 - Longhorn version: v1.3.2 (Harvester v1.1.0)
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl):

## Additional context

Thanks for @ibrokethecloud's reporting!

**Comment by roger-ryao** [comment]
Hi @innobead 
    I will verify this issue after @weizhe0422 fixes it.

**Comment by weizhe0422** [comment]
Update the e2e result for backingimage is pass
![image](https://user-images.githubusercontent.com/17548901/206347161-bff9b954-1dc9-4927-8d03-fefddd2aea4d.png)

**Comment by roger-ryao** [comment]
**Verified on master-head 20221208**

- longhorn master-head ([13ac2a66](https://github.com/longhorn/longhorn/commit/13ac2a66415d518218bacd9e15528b32549db89d)) 
- longhorn-manager master-head ([5a07e68](https://github.com/longhorn/longhorn-manager/commit/5a07e68388a04d1d6e2536b44d4f02c2b6dc4030))


**The test steps**

Scenario 1
[Ref #4865 (comment)](https://github.com/longhorn/longhorn/issues/4865#issuecomment-1340669204)


Scenario 2
1. Upload a local file as a backing image and check the `BackingImageDatasource` POD should be launched
2. Access to `BackingImageDatasource` POD's node 
3. Find out the process ID and, freeze the process
`ps aux | grep 'backing-image-manager --debug data-source' |grep -v grep | awk '{print $2}' | xargs -i kill -STOP {}`


**Result** 
Scenario 1
1. Unblock `BackingImageDatasource` pod ingress and outgress package
2. The backing image downloading will be continued and marked as ready
3. 
![Screenshot_20221208_181515](https://user-images.githubusercontent.com/104337648/206420848-c7198835-4d55-4f1b-ae36-b5e0f88f8a67.png)



Scenario 2
1. Unfreeze the process
`ps aux | grep 'backing-image-manager --debug data-source' |grep -v grep | awk '{print $2}' | xargs -i kill -CONT {}`
2. The backing image downloading will be marked as failed without any panic
![Screenshot_20221208_175823](https://user-images.githubusercontent.com/104337648/206419479-9e037155-cc40-48d7-aa63-f779c576c7bb.png)

**Comment by innobead** [comment]
@weizhe0422 What longhorn version is the user using? This fix is only going to 1.4.0 and 1.3.3.

**Comment by w13915984028** [comment]
@innobead 
The issue is reported in Harvester v1.1.1, which has longhorn: helm.sh/chart: longhorn-1.3.2.

This fix, #4865, should not in Harvester v1.1.1. 

And futher, will big file causes potential issue to LH? That's the reason I want to know the ever tested largest single file size. thanks.

**Comment by innobead** [comment]
I see. If that's streaming, don't think that's the problem. Let's see the update from @roger-ryao / @shuo-wu .

**Comment by roger-ryao** [comment]
> @roger-ryao @shuo-wu : what is the largest file ever tested?
> 
> users are reporting with a 75GB local file to upload, it failed. [harvester/harvester#3450](https://github.com/harvester/harvester/issues/3450)
> 
> could it be related to POD resource restriction or disk space pressure ? thanks. cc @chrisho

Hi @w13915984028 
I verified this issue with a 5GB local file to upload.

**Comment by w13915984028** [comment]
@roger-ryao @innobead  

We suspect the large image file (e.g. 50Gb ~ 100Gb) may cause pressue to the PODs in the uploading path, when possible, please test with such big image files. And also add them into auto test.

LH will have a specification, saying which size of image can be handled smoothly.

thanks.

**Comment by shuo-wu** [comment]
@w13915984028 Is the large file uploading failure caused by this kind of issue?
https://github.com/longhorn/longhorn/issues/4902

**Comment by w13915984028** [comment]
@shuo-wu 
The straightforward issue is 
https://github.com/harvester/harvester/issues/3450,  some POD crashed inbetween the uploading an 75Gb single-file image.

https://github.com/longhorn/longhorn/issues/4902 touchs more components, the vm-importer is an upper layer controller, which utilize the uploading functionality.

**Comment by innobead** [comment]
@w13915984028 Let's create another issue to track large file uploading instead to clarify this further.

**Comment by w13915984028** [comment]
OK, let discuss further in https://github.com/longhorn/longhorn/issues/5395, thanks.
